src/index.ts

```diff
@@ -0,0 +1,88 @@
+import crypto from 'crypto';
+import { Request, RequestHandler, NextFunction } from 'express';
+
+import stringToBuffer from './stringToBuffer';
+import validateArguments from './validateArguments';
+import AuthError from './errors';
+
+export type Options = {
+    algorithm: string;
+    identifier: string;
+    header: string;
+    maxInterval: number;
+    minInterval: number;
+}
+
+const defaults: Options = {
+    algorithm: 'sha256',
+    identifier: 'HMAC',
+    header: 'authentication',
+    maxInterval: 60 * 5,
+    minInterval: 0 
+}
+
+export { AuthError };
+
+export function HMAC(secret: string, options: Partial<Options> = {}): RequestHandler {
+    const mergedOpts: Options = { ...defaults, ...options };
+
+    const error: TypeError | void = validateArguments(secret, mergedOpts);
+    if (error instanceof TypeError) {
+        throw error;
+    }
+
+    return function(request: Request, _, next: NextFunction): void {
+        if (!(mergedOpts.header in request.headers)) {
+            return next(new AuthError(`Header provided not in sent headers. Expected ${mergedOpts.header} but not found in request.headers`));
+        }
+
+        // see https://github.com/microsoft/TypeScript/issues/41612
+        // for why we can't do request.headers[options.header] in the conditional
+        const header: string | string[] | undefined = request.headers[mergedOpts.header];
```

Comment:
Typescript can infer from the return type of  `request.headers[mergedOpts.header]` what the type of `header` will be, so unlike some other typed languages you don't need to be this verbose - in practice we generally let typescript and the IDE do a lot of the work for us. The same goes throughout this file where you're declaring new constants - you can skip redeclaring the types as it just adds more noise to the file.